From: Liam Howlett <liam.howlett@oracle.com>
Subject: maple_tree: fix mas_empty_area_rev() search exhaustion at root node
Date: Tue, 12 Jul 2022 01:50:36 +0000

When the possibilities of finding a gap are exhausted at the root node,
the error code was not returned which results in returning a range that
is currently occupied.  Add the necessary check for the error code and
return to the caller.

Link: https://lkml.kernel.org/r/20220712014957.4064879-1-Liam.Howlett@oracle.com
Fixes: d0aac5e48048 (Maple Tree: add new data structure)
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Reported-by: Alexander Gordeev <agordeev@linux.ibm.com>
Tested-by: Alexander Gordeev <agordeev@linux.ibm.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/maple_tree.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/lib/maple_tree.c~maple-tree-add-new-data-structure-fix-5
+++ a/lib/maple_tree.c
@@ -5287,6 +5287,9 @@ int mas_empty_area_rev(struct ma_state *
 		}
 	}
 
+	if (mas_is_err(mas))
+		return xa_err(mas->node);
+
 	if (unlikely(mas->offset == MAPLE_NODE_SLOTS))
 		return -EBUSY;
 
_
